garrysmod/lua/vgui/dscrollpanel.lua

```diff
@@ -27,119 +30,106 @@ function PANEL:Init()
 
 function PANEL:OnChildAdded( child )
-
 	self:AddItem( child )
-
 end
 
 function PANEL:SizeToContents()
-
-	self:SetSize( self.pnlCanvas:GetSize() )
-
+	local canvas = self:GetCanvas()
+	if IsValid( canvas ) then
+		self:SetSize( canvas:GetSize() )
+	end
 end
 
 function PANEL:GetVBar()
-
 	return self.VBar
-
 end
 
 function PANEL:GetCanvas()
-
 	return self.pnlCanvas
-
 end
 
 function PANEL:InnerWidth()
-
 	return self:GetCanvas():GetWide()
-
-end
-
-function PANEL:Rebuild()
-
-	self:GetCanvas():SizeToChildren( false, true )
-
-	-- Although this behaviour isn't exactly implied, center vertically too
-	if ( self.m_bNoSizing && self:GetCanvas():GetTall() < self:GetTall() ) then
-
-		self:GetCanvas():SetPos( 0, ( self:GetTall() - self:GetCanvas():GetTall() ) * 0.5 )
-
-	end
-
 end
 
 function PANEL:OnMouseWheeled( dlta )
-
-	return self.VBar:OnMouseWheeled( dlta )
-
+	local vbar = self:GetVBar()
+	if IsValid( vbar ) then
+		return vbar:OnMouseWheeled( dlta )
+	end
 end
 
 function PANEL:OnVScroll( iOffset )
-
-	self.pnlCanvas:SetPos( 0, iOffset )
-
+	local canvas = self:GetCanvas()
+	if IsValid( canvas ) then
+		canvas:SetPos( 0, iOffset )
+	end
 end
 
 function PANEL:ScrollToChild( panel )
-
 	self:InvalidateLayout( true )
 
-	local x, y = self.pnlCanvas:GetChildPosition( panel )
-	local w, h = panel:GetSize()
-
-	y = y + h * 0.5
-	y = y - self:GetTall() * 0.5
-
-	self.VBar:AnimateTo( y, 0.5, 0, 0.5 )
+	local canvas = self:GetCanvas()
+	if IsValid( canvas ) then
+		local vbar = self:GetVBar()
+		if IsValid( vbar ) then
+			local _, y = canvas:GetChildPosition( panel )
+			vbar:AnimateTo( (y + panel:GetTall() / 2) - self:GetTall() / 2, 0.5, 0, 0.5 )
+		end
+	end
+end
 
+function PANEL:Rebuild()
+	local canvas = self:GetCanvas()
+	if IsValid( canvas ) then
+		canvas:SizeToChildren( false, true )
+
+		-- Although this behaviour isn't exactly implied, center vertically too
+		local canvasHeight = canvas:GetTall()
+		if self.m_bNoSizing and (canvasHeight < self:GetTall()) then
+			canvas:SetPos( 0, (self:GetTall() - canvasHeight) / 2 )
+		end
+	end
 end
 
--- Avoid an infinite loop
 function PANEL:PerformLayoutInternal()
-
-	local Tall = self.pnlCanvas:GetTall()
-	local Wide = self:GetWide()
-	local YPos = 0
-
-	self:Rebuild()
-
-	self.VBar:SetUp( self:GetTall(), self.pnlCanvas:GetTall() )
-	YPos = self.VBar:GetOffset()
-
-	if ( self.VBar.Enabled ) then Wide = Wide - self.VBar:GetWide() end
-
-	self.pnlCanvas:SetPos( 0, YPos )
-	self.pnlCanvas:SetWide( Wide )
-
-	self:Rebuild()
-
-	if ( Tall != self.pnlCanvas:GetTall() ) then
-		self.VBar:SetScroll( self.VBar:GetScroll() ) -- Make sure we are not too far down!
+	local canvas = self:GetCanvas()
+	if IsValid( canvas ) then
+		local canvasHeight = canvas:GetTall()
+		local width = self:GetWide()
+
+		self:Rebuild()
+
+		local vbar = self:GetVBar()
+		if IsValid( vbar ) then
+			local height = self:GetTall()
+			vbar:SetUp( height, canvasHeight )
+			if vbar.Enabled and (canvasHeight < height) then

```

Comment:
This check will almost never pass as `vbar.Enabled` is set to `canvasHeight > height`, resulting in bugged spacing:

![image](https://user-images.githubusercontent.com/3299036/212738114-50bde7b0-2ef5-4471-914c-c39b0754d685.png)

(Note the text entry going behind the scrollbar)

Reply:
Hmmm, I must have missed it somewhere, I'll try to fix it on this branch